### ISSUE TITLE
EINVAL error when processing xattrs (borg 1.2.1)

### ISSUE THREAD

**Issue by Rudd-O** [borgbackup/borg#6988]
Some files on my system, when they are being backed up, cause `borg` to emit this:

```
Aug 23 02:37:18 penny.dragonfear borg-offsite-backup[937998]: srv/shared/Albums/Bolitas de agua saltan/PXL_20220324_124712302.mp4: extended stat: [Errno 22] Invalid argument: '<FD 12>'
```

This causes the backup to finish with error code 1.

Of note is that the files have extended attributes:

```
[root@penny /]# getfattr 'srv/shared/Albums/Bolitas de agua saltan/PXL_20220324_124712302.mp4'
# file: srv/shared/Albums/Bolitas de agua saltan/PXL_20220324_124712302.mp4
user.DOSATTRIB
user.security.selinux
```

This did not appear to happen with `borg` 1.2.0.

I use Fedora 36 and ZFS.

**Comment by ThomasWaldmann** [comment]
Strange, that "<FD 12>" means file descriptor 12 (== an open file) and not sure what could be invalid with that.

Also not sure where exactly the exception comes from, all these are done under "extended stat":
- ("bsd" / fs)flags
- xattrs
- ACLs

Do you run borg as root?

**Comment by ThomasWaldmann** [comment]
@Rudd-O see #6990 for a code change that will make it more clear where such an error comes from.

If you can reproduce, could you please find the minimum set of options that makes the error go away for you?
- `--noflags`
- `--noxattrs`
- `--noacls`
- (a combination of the above)

**Comment by Rudd-O** [comment]
@ThomasWaldmann indeed I run borg as root.  I can repro every time.

Let me test those options now.

**Comment by Rudd-O** [comment]
`--noxattrs` made it go away.  I did not try the other ones because I know the erroring files don't have ACLs, or flags (that's a BSD thing IIRC).

Of course, I would prefer the extended attributes to be saved.

**Comment by ThomasWaldmann** [comment]
OK, so we now know for sure it's something with listing or getting the xattrs.

Is it always the same file(s) or do the files it warns about (in that way) change each run?

Do the failing files have xattr **keys/names** that files it successfully backs up do not have?

Do the failing files have xattr **values** that files it successfully backs up do not have?

**Comment by ThomasWaldmann** [comment]
Oh, I just noticed that I maybe read this wrong:
```
.../PXL_20220324_124712302.mp4: extended stat: [Errno 22] Invalid argument: '<FD 12>'
```
- The filename on the left is output by borg (borg know which filename it is processing at that moment).

- `extended stat` also comes from borg (`backup_io` exception handling wrapper).

- `[Errno 22] Invalid argument: '<FD 12>'` that part comes from the OS. I maybe misinterpreted it that it was saying that the file descriptor is invalid. But I guess it rather wants to tell us that **some argument (not the fd) of the call was invalid** and additionally tells us the filename/fd the call operated on was fd 12.

**Comment by ThomasWaldmann** [comment]
- getxattr man page does not list EINVAL for getxattr, but: `In addition, the errors documented in [stat(2)](https://man7.org/linux/man-pages/man2/stat.2.html) can also occur.`
- listxattr man page does not list EINVAL for listxattr, but `In addition, the errors documented in [stat(2)](https://man7.org/linux/man-pages/man2/stat.2.html) can also occur.`
- stat man page tells: `EINVAL (fstatat()) Invalid flag specified in flags.` (flags being stuff like AT_SYMLINK_NOFOLLOW)

**Comment by ThomasWaldmann** [comment]
reading there https://github.com/torvalds/linux/blob/master/fs/f2fs/xattr.c gives:
- EINVAL in case getxattr / setxattr is called with name==NULL
- EINVAL for unexpected handler (not user, security, trusted)

https://github.com/torvalds/linux/blob/master/fs/ext4/xattr.c gives:
- EINVAL for some fs error condition (EA inode %lu does not have EXT4_EA_INODE_FL flag)

https://github.com/openzfs/openzfs/blob/master/usr/src/uts/common/fs/xattr.c
- ???

**Comment by ThomasWaldmann** [comment]
`user.security.selinux` is that a valid xattr name? shouldn't that be `security.selinux`?

**Comment by ThomasWaldmann** [comment]
Guess that could be it (see previous comment):
https://github.com/openzfs/zfs/blob/9681de4657686d0ed19ca18d578513e74395f00f/include/sys/fs/zfs.h#L1794

**Comment by ThomasWaldmann** [comment]
@Rudd-O can you try the code from #6990 (and use `--list` option and not `--progress` for borg create)?

I have some hope that this code fixes your issue (not the one that you might have an invalid xattr name in your fs though).

**Comment by ThomasWaldmann** [comment]
While the root cause is likely the invalid xattr name given by `listxattr`, this is also a bug in borg's error handling:
- borg must expect invalid xattr names returned by `listxattr`
- borg should not stop processing xattrs if just one of them is somehow invalid / can't be read, but continue processing the remaining xattrs
- borg should not stop processing the whole file due to such an error in xattr processing

**Comment by Rudd-O** [comment]
> OK, so we now know for sure it's something with listing or getting the xattrs.

Yep.

> Is it always the same file(s) or do the files it warns about (in that way) change each run?

Always same files.
 
> Do the failing files have xattr **keys/names** that files it successfully backs up do not have?

I couldn't tell you right this very moment, I'm on a trip.

> Do the failing files have xattr **values** that files it successfully backs up do not have?

Same. Will have to postpone it to next week.

What commands should I run to produce a data dump that would allow you to repro locally?  Alternative, which files should I add `printf()`s to?

**Comment by ThomasWaldmann** [comment]
I suspect that xattr name `user.security.selinux` is the root cause of this issue.

Maybe some older zfs version still allowed this name and current zfs version does not. If that is the case, there might be no easy way to reproduce this via the filesystem.

**Comment by Rudd-O** [comment]
Many files successfully backed up have that attribute, so I don't think that's the case.  I'm more thinking perhaps it's the DOSATTRIB thing.  `user.security.selinux` is, as far as I know, a standard and valid attribute name.

**Comment by Rudd-O** [comment]
Would still be useful to know this:

>  What commands should I run to produce a data dump that would allow you to repro locally? Alternative, which files should I add printf()s to?

**Comment by ThomasWaldmann** [comment]
If I understand the code correctly, the 1st link in https://github.com/borgbackup/borg/issues/6988#issuecomment-1225372972 specifically invalidates stuff like `user.security.selinux`.

In that case "user" would be the namespace and "security.selinux" would be the xattr name.
That code makes sure that an xattr name does **not** start with a namespace name (like "security.").

Also, IIRC this needs to be `security.selinux` (without `user.`).

**Comment by ThomasWaldmann** [comment]
You could also try using xattr commands to remove the one or the other xattr and check if that fixes the issue.

Before you do that, it would be very useful to try the code in #6990 - that should just ignore troublesome xattrs.

**Comment by Rudd-O** [comment]
Reminder to myself: check out #6990 .

**Comment by Rudd-O** [comment]
Guess who is about to give you more error data!

That's right!

OK, so I applied the patch you referenced, and here is the output:

```
when getting extended attribute user.security.selinux: [Errno 22] Invalid argument: '<FD 6>'                                                                                                                      
```

So yes, it's the SELinux attribute in question.

When i retrieve the attribute by hand:

```
[root@penny Bolitas de agua saltan]# getfattr -d *
# file: PXL_20220324_124647555.mp4
user.DOSATTRIB=0sAAAEAAQAAABRAAAAIAAAAIfok7yzX9gBh+iTvLNf2AE=
PXL_20220324_124647555.mp4: user.security.selinux: Invalid argument

# file: PXL_20220324_124712302.mp4
user.DOSATTRIB=0sAAAEAAQAAABRAAAAIAAAAA6yqbuzX9gBDrKpu7Nf2AE=
PXL_20220324_124712302.mp4: user.security.selinux: Invalid argument
```

**Comment by Rudd-O** [comment]
I have no clue what is adding those stupid `user.security.selinux` attributes.  Perhaps it's Syncthing?  AAAAAAAA!

**Comment by Rudd-O** [comment]
I can't even remove the stupid attributes!

**Comment by Rudd-O** [comment]
How tf is this even possible??

```
[root@penny srv]# attr -l ./shared/Knowledge/Unsorted/Images/SlimSocial-47.jpg
Attribute "selinux" has a 35 byte value for ./shared/Knowledge/Unsorted/Images/SlimSocial-47.jpg
Attribute "DOSATTRIB" has a 32 byte value for ./shared/Knowledge/Unsorted/Images/SlimSocial-47.jpg
Attribute "security.selinux" has a -1 byte value for ./shared/Knowledge/Unsorted/Images/SlimSocial-47.jpg
```

**Comment by ThomasWaldmann** [comment]
I suspect these should be `security.selinux` (in the `security` namespace, not in the `user` namespace).

OK, so when you tried borg with the patch, it logged that warning you posted above, but it continued to backup up **that file** (contents, all other xattrs) and all other files after that, right? that's the expected behaviour and the intention of this PR.

**Comment by ThomasWaldmann** [comment]
My theory is that an older version of zfs allowed to set this attribute, then they added this additional check and now it does not let you access it any more.

**Comment by Rudd-O** [comment]
Yep, gotta file a ticket with these folks.  Thanks.

**Comment by ThomasWaldmann** [comment]
Let's limit the scope of this issue to the bad error handling borg showed here:
- not specific enough error messages
- cancelling the backup of all remaining xattrs
- cancelling the backup of the whole file

I consider these all solved by #6990 (1.2-maint branch), so we can close this after forward port to master.

Some related (but AFAIK freebsd specific issue) was split off into #6997.

**Comment by ThomasWaldmann** [comment]
Fixed by #7008
